### ISSUE TITLE
Parallelization of infer.upperlimit

### ISSUE THREAD

**Issue by matthewfeickert** [scikit-hep/pyhf#1327]
# Description

Just starting this so that we don't loose it to the mist of times: During the [Reinterpretation Forum tutorial](https://pyhf.github.io/tutorial-Reinterpretation-Forum-2021/) @jackaraz @WolfgangWaltenberger had good questions on if there was a way to improve the speed of the `upperlimit` API, as it currently performs a POI scan

https://github.com/scikit-hep/pyhf/blob/aab6953d716daeaf1e1660bed35e5b7dc0363580/src/pyhf/infer/intervals.py#L52-L55

They had suggested [Brent bracketing](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.brentq.html), which I don't have much experience with using yet. 

@jackaraz @WolfgangWaltenberger I think you both mentioned that you have code that uses this at the moment. Can you link to that here so we can take a look at what you're currently doing? 

@WolfgangWaltenberger had mentioned there were parallelization issues with function bracketing (if I remember right), but for a POI scan it is definitely possible to parallelize, but we'll need to look at how to do so across the backends.

**Comment by kratsg** [comment]
See #1274

**Comment by jackaraz** [comment]
Hi @matthewfeickert, thanks for opening the thread. The code is in [this link](https://bazaar.launchpad.net/~ma5dev/madanalysis5/ma5_pyhf/view/head:/madanalysis/misc/run_recast.py#L1650), but in case it's messy, let me explain what is going on there.

After MadAnalysis calculates all the SR yields, we create a signal patch as suggested on your website. For instance, an example patch that is written by the code is as follows;

```
[{'op': 'add',
  'path': '/channels/0/samples/0',
  'value': {'name': 'MA5_signal',
   'data': [0.0, 0.0, 5.5428],
   'modifiers': [{'data': None, 'name': 'lumi', 'type': 'lumi'},
    {'data': None, 'name': 'mu_SIG', 'type': 'normfactor'}]}},
 {'op': 'remove',
  'path': '/channels/1/samples/0',
  'value': {'name': 'MA5_signal', 'data': [], 'modifiers': []}},
 {'op': 'remove',
  'path': '/channels/2/samples/0',
  'value': {'name': 'MA5_signal', 'data': [], 'modifiers': []}}]
```

This will be patched to ATLAS-SUSY-2018-031 analysis JSON files, specifically, let's say, the one for Region A. To get observed and expected CLs, I create a workspace and calculate it as follows;

```python
workspace = pyhf.Workspace(regA_bkg) # this is the json file provided by the colaboration
model     = workspace.model(patches=[regA_sig], # this is the patch given above
                            modifier_settings={'normsys': {'interpcode': 'code4'},
                                               'histosys': {'interpcode': 'code4p'}})
CLs_obs, CLs_exp = pyhf.infer.hypotest(1., 
                          workspace.data(model),
                          model, 
                          test_stat="qtilde",
                          par_bounds=model.config.suggested_bounds(),
                          return_expected=True)
```

So this gives me two CLs value for expected and observed. I also need to calculate expected (observed) excluded cross-section at 95% CL (which is happening [here](https://bazaar.launchpad.net/~ma5dev/madanalysis5/ma5_pyhf/view/head:/madanalysis/misc/run_recast.py#L1419), but I'll explain). For that, I'm varying the xsec value of my theory and calculating the exclusion for different xsec values on the same patch that I provided above. but of course, pyhf can't calculate the CLs for every single xsec value, so I expand the parameter bounds for 3 iterations

```python
update_bounds[model.config.poi_index] = (0,2*update_bounds[model.config.poi_index][1])
```

And if it still can't calculate CLs, I say that this xsec value is excluded. So I'm scanning xsec values from 0.1 to 10 to be able to find the exclusion xsec above, and below 0.95 then I try to minimize the scan and get the exclusion cross-section (done [here](https://bazaar.launchpad.net/~ma5dev/madanalysis5/ma5_pyhf/view/head:/madanalysis/misc/run_recast.py#L1455)). Does this makes sense or do you think I'm trying to kill a fly with a bazooka? It feels like by scanning mu, I can find the exclusion xsec more efficiently.

**Comment by WolfgangWaltenberger** [comment]
So to add, one common theme is  that we wish to find 95% CL limits, i.e. the root of f(mu) = CLs(mu) - 0.95.  Your code in the tutorial solved it by scanning for mu. In SModelS and in MA5 (Jack's link at line 1376) we perform a bracketing technique (probably doesn't matter so much if brent or toms748). So obviously bracketing needs fewer evaluations for CLs(mu) for the same accuracy, but you can't really parallelize it, because the n-th iteration depends on the outcome of the (n-1)-th iteration. So perhaps the best solution is  a kind of mix of parallelized scanning and bracketing -- also depending on the hardware platform one is on. [If you dont parallelize, bracketing is clearly better]. Strictly speaking, this computation does not need to be inside of pyhf, after all we are doing it outside right now. But pyhf would IMHO yes be the best place for this functionality, as we would be using it frequently, it is a very common and typical use case.

**Comment by kratsg** [comment]
Hi @WolfgangWaltenberger , can you describe how you're parallelizing it with a naive mu-scan? Do you simply just take the 50 points to evaluate hypotest at, and naively parallelize/multiprocess/multithread?

I guess you're looking for `concurrent.futures` support?

**Comment by WolfgangWaltenberger** [comment]
Ah @kratsg sorry that might be a misunderstanding. We're currently not parallelizing that step at all. We only use brentq. What I meant to say is that a scan for mu seems straightforwardly parallelizable, yes using concurrent.futures. And I wondered, if one couldnt e.g. evaluate in parallel at, say 10 equidistant values of mu. Then choose those two subsequent values where the sign flipped [ as we are looking for the root of f(mu) ], and evaluate again in parallel at 10 values of mu within that interval. That way you obtain a fast yet precise value of mu. But you guys obviously know much better how this all fits into pyhf overall. There may be a smarter way to do this.

**Comment by kratsg** [comment]
Fair. I'd say that parallelization versus auto-scans (e.g. TOMS or brent bracketing) are two separate things. E.G. a defined algorithm should be able to describe how to run for a provided executor. Would `SModelS` prefer brent bracketing or parallelization? Or is the preference just for being able to get a faster, accurate upper limit regardless?

**Comment by WolfgangWaltenberger** [comment]
Yes, there is no preference from the SModelS side. Obviously we want the most accurate result in the fastest possible way :)  
but as always we are willing to sacrifice accuracy for speed. So far we used brent because it was so easy and simple to implement.
